pennylane/transforms/dynamic_one_shot.py

```diff
@@ -113,6 +119,29 @@ def processing_fn(results, has_partitioned_shots=None):
     return output_tapes, processing_fn
 
 
+@dynamic_one_shot.custom_qnode_transform
```

Comment:
So I'm confused about even allowing this to be applied to a qnode.  If we are using it by default in all situations it could be used, should we even be allowed to apply it a second time?

Reply:
This change was [suggested](https://github.com/PennyLaneAI/pennylane/pull/5317#pullrequestreview-1920899407) by @mudit2812 as a way to raise early if a device didn't have `supports_mid_measure`. 

What I would like is apply `dynamic_one_shot` by default on QNodes:

- with legacy devices declaring `supports_mid_measure` in their `capabilities`
- a tape executed with finite shots and MCMs (and without parameter broadcasting)

Given a QNode, it always possible a user will apply `dynamic_one_shot` on top of the default, or maybe just 
`dynamic_one_shot(dynamic_one_shot(circuit))`, which is why we needed `prune_dynamic_transform`. 

Is your suggestion to move the logic [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/_device.py#L721)?

Reply:
Yes. Moving into `Device.batch_transform`. That is the location for devices to opt- into any batch transforms.